### ISSUE TITLE
The validation dataset is not actuallly used for validation?

### ISSUE THREAD

**Issue by YunYunY** [SakurajimaMaiii/TSD#5]
Hi,
Thank you for sharing your work. I found there might be an error in the code. Please correct me if I'm wrong. 

In the paper Section 4 Implementation, you claimed ` We split all images from the source
domains to 80%/20% for training and validation.` But in the code ``train.py save_checkpoint('model.pkl', algorithm, args) `` the trained model is saved no matter wheher the condition ``acc_record['valid'] > best_valid_acc`` is satisfied or not.

**Comment by SakurajimaMaiii** [comment]
Thanks for your question. __I think you are right.__ The model selection has some bugs.
 I ignore the fact that I only save the model after training (not the best model on the validation set).   
So, in fact, the model for TTA methods we use is the last epoch model not the best model on the validation set.
__The accuracy of ERM  reported in the paper is the true target acc using training domain validation.__ (We get it from `done.txt` after training)
However, the pre-trained model TTA methods use is the last epoch model.  So, the performance improvement is not true. 
Although I think the model from the last epoch performs similarly to the best model on the validation set, the results need to be updated.
We can confirm:
1. The comparison among all TTA methods is __fair__ because we use the same pre-trained model (model after training) on source domains for all methods.
2. The relative improvement needs to be re-calculated. But the rank of different methods should be the same.

__Thanks for pointing out this question. I will update the code and new results as soon as possible.__

**Comment by SakurajimaMaiii** [comment]
If you want to get the ERM accuracy of saved model (model after training), you can use `ERM` in [adapt_algorithm.py](https://github.com/SakurajimaMaiii/TSD/blob/master/code/adapt_algorithm.py)
```
python unsupervise_adapt.py --dataset PACS\
                            --data_dir your_data_dir\
                            --adapt_alg ERM\ 
                            --pretrain_dir your_pretrain_model_dir\
```
This commond can get the ERM accuracy of the model used for TTA.  
After this, feel free to use different TTA methods to check the pefromance after test time adaptation.

**Comment by YunYunY** [comment]
Thank you for the quick response. I agree that even the save model is not the **best**, it's still a fair comparison for different TTAs.

**Comment by SakurajimaMaiii** [comment]
Hi I have fixed this bug.
Please see commit `f300fe9`.
